### PR TITLE
Update EVM account mapping, add contract account handling

### PR DESCRIPTION
**Pull request by Mr-Leshiy**

### CODE REVIEW

**Review by Zeegomo — CHANGES_REQUESTED** [review]
The `AccountState` thing looks a bit like a state machine to me (not mapped, mapped, contract account, ...), can we make it more explicit?

File: chain-impl-mockchain/src/ledger/evm.rs [hunk 1]
```diff
@@ -45,8 +57,11 @@ impl AddressMapping {
         }
     }
 
-    fn jor_address(&self, evm_id: &EvmAddress) -> Option<&JorAddress> {
-        self.evm_to_jor.lookup(evm_id)
+    fn jor_address(&self, evm_id: &EvmAddress) -> JorAddress {
+        match self.evm_to_jor.lookup(evm_id).cloned() {
+            Some(jor_address) => jor_address,
+            None => transfrom_evm_to_jor(evm_id),
```
[the remark this answers is not in the document]
Comment (Mr-Leshiy) [reply chain-impl-mockchain/src/ledger/evm.rs:66]: fixed

File: chain-impl-mockchain/src/accounting/account/mod.rs [hunk 2]
```diff
@@ -326,8 +357,113 @@ mod tests {
...
+
+        assert_eq!(
+            ledger.get_state(&account_id2),
+            Err(LedgerError::NonExistent)
+        );
+
+        assert_eq!(
+            ledger.get_state(&account_id3),
+            Err(LedgerError::NonExistent)
+        );
```
[the remark this answers is not in the document]
Comment (Mr-Leshiy) [reply chain-impl-mockchain/src/accounting/account/mod.rs:388]: Do you mean for the `account_id3` ? Or do you mean that all these checks are not necessary ?
 ```      
 assert_eq!(
            ledger.get_state(&account_id1),
            Err(LedgerError::NonExistent)
        );

        assert_eq!(
            ledger.get_state(&account_id2),
            Err(LedgerError::NonExistent)
        );

        assert_eq!(
            ledger.get_state(&account_id3),
            Err(LedgerError::NonExistent)
        );
```

[the remark this answers is not in the document]
Comment (Zeegomo) [reply chain-impl-mockchain/src/accounting/account/mod.rs:388]: all of those

File: chain-impl-mockchain/src/ledger/evm.rs [hunk 3]
```diff
@@ -56,23 +71,32 @@ impl AddressMapping {
...
             .map_err(|_| Error::ExistingMapping(jor_id.clone(), evm_id))?;
         let jor_to_evm = self
             .jor_to_evm
             .insert(jor_id.clone(), evm_id)
-            .map_err(|_| Error::ExistingMapping(jor_id, evm_id))?;
+            .map_err(|_| Error::ExistingMapping(jor_id.clone(), evm_id))?;
+
+        // should update and move account evm account state
+        let old_jor_id = transfrom_evm_to_jor(&evm_id);
+        accounts = accounts.evm_move_state(jor_id, &old_jor_id)?;
```
[the remark this answers is not in the document]
Comment (Mr-Leshiy) [reply chain-impl-mockchain/src/ledger/evm.rs:94]: Generally no, we only allow to map to that jormungandr account which user owns, signature validation comes in the another place.

**Comment by Mr-Leshiy** [comment]
@Zeegomo 
>The AccountState thing looks a bit like a state machine to me (not mapped, mapped, contract account, ...), maybe we make it more explicit?
Yes, good point, but need to think how it is better to do it.

File: chain-impl-mockchain/src/ledger/evm.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Zeegomo) [reply chain-impl-mockchain/src/ledger/evm.rs:94]: Yes, but that is currently allowed. I would have expected the implementation to progress in the opposite order not to leave exploitable gaps, but that's not too important in the end, just do not forget to add a ticket for that if it does not exist yet

[the remark this answers is not in the document]
Comment (Mr-Leshiy) [reply chain-impl-mockchain/src/ledger/evm.rs:94]: Sure, we have already tickets for that :)

File: chain-impl-mockchain/src/ledger/evm.rs [hunk 4]
```diff
@@ -42,6 +42,16 @@ pub struct AddressMapping {
     jor_to_evm: Hamt<DefaultHasher, JorAddress, EvmAddress>,
 }
 
+/// One way transforming procedure from `EvmAddress` to `JorAddress`.
+/// This allows to map any `EvmAddress` to the `JorAddress` before explicit execution of the `EvmMapping` transaction.
+/// Intention - is to have possibility to map EVM Contarct accounts with the Jormungandr accounts.
```
Comment (Zeegomo) [inline chain-impl-mockchain/src/ledger/evm.rs:47]: This is confusing though, contract accounts should never be part of a `EvmMapping` transaction AFAIK, right?

Comment (Mr-Leshiy) [reply chain-impl-mockchain/src/ledger/evm.rs:47]: Yes, contract account should be a part of the `EvmMapping` certificate, will try to paraphrase it. I meant that with the usage of this function we can link contract account with the jormungandr.

Comment (Zeegomo) [reply chain-impl-mockchain/src/ledger/evm.rs]: > contract account should be a part of the EvmMapping certificate

How can this happen?

Comment (Mr-Leshiy) [reply chain-impl-mockchain/src/ledger/evm.rs]: Sorry, mistaken 😓
> contract account should **NOT** be a part of the EvmMapping certificate

File: chain-impl-mockchain/src/ledger/evm.rs [hunk 5]
```diff
@@ -21,21 +24,43 @@ use thiserror::Error;
...
-#[derive(Clone, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub struct AddressMapping {
     evm_to_jor: Hamt<DefaultHasher, EvmAddress, JorAddress>,
     jor_to_evm: Hamt<DefaultHasher, JorAddress, EvmAddress>,
 }
 
+/// One way transforming procedure from `EvmAddress` to `JorAddress`.
+/// This allows to map any `EvmAddress` to the `JorAddress` before explicit execution of the `EvmMapping` transaction.
+/// Intention - is to have possibility to link an EVM Contarct account with a Jormungandr account.
```
Comment (Zeegomo) [inline chain-impl-mockchain/src/ledger/evm.rs:46]: those are two slightly different use cases.
1) standard account not yet mapped, but that _could_ be mapped in the future
2) contract account that can't possibly be mapped

File: chain-impl-mockchain/src/ledger/evm.rs [hunk 6]
```diff
@@ -21,21 +24,43 @@ use thiserror::Error;
...
+///
+/// Algorithm description:
+///  1. Get `evm_address` bytes representation -> evm_address_bytes
+///  2. Append b"evm" bytes prefix to the evm_address_bytes -> bytes_data
```
Comment (Zeegomo) [inline chain-impl-mockchain/src/ledger/evm.rs:50]: nit: append a prefix = prepend

File: chain-impl-mockchain/src/ledger/evm.rs [hunk 7]
```diff
@@ -21,21 +24,43 @@ use thiserror::Error;
...
+///  3. Calculate hash (pub struct Hash(crypto::Blake2b256) hash) from the bytes_data -> hash_bytes
```
Comment (Zeegomo) [inline chain-impl-mockchain/src/ledger/evm.rs:51]: nit: don't need a reference to the specific Rust impl here, just mentioning blake2b256 is enough

**Review by saibatizoku — APPROVED** [review]

**Pull request merged**